Bearded.UI/Controls/implementations/ListControl.cs

```diff
@@ -229,6 +235,9 @@ public void Reload()
 
         private void calculateTotalHeight()
         {
+            if (itemSource == null)
+                throw new InvalidOperationException("The ItemSource is not set.");
```

Comment:
This method is only called from a place where we already verified this isn't null. Do we have to check it again?

Reply:
Answering your question here, as on a private method and internal method one might have different opinions, though I think an internal method should have the same checks as a public method should, but that's my opinion.


The C# compiler does not know that itemSource is or isn't null here, so it assumes it can be null, and thus gives an error (or warning). I don't know if the compiler doesn't know this on purpose, or because it can't evaluate that deeply.

You can solve this error by either checking for null and then throwing an exception, or by using the null-forgiving operator (`!`). I don't know if a Debug.Assert works, but I think not.
The pro of using the `!` is that it doesn't make your code ugly by these two throwing lines (and that your code is a tiiiny bit faster, for it doesn't have to evaluate an if that always has the same outcome).
The con is that if you ever call this method from another place, and you forget to check for null, the compiler will not warn you about it. And the fact that you have nullable errors turned on, and the compiler not warning you about it may give you a false sense of security causing you to forget it.

The other alternative I guess is to not treat the nullable warnings as errors, and just leave the warning there and ignore it. Personally I think that teaching yourself to ignore warnings is a really bad choice though.


I don't mind using the `!` here, it's your project after all, but then it's on you (or Tom I suppose) to be careful even when your compiler tells you all is well.

Reply:
How about passing in the verified-non-null collection into the private methods?

That way we get compile time safety (and I fully agree, we should not ignore these warnings, and we should in fact make them errors!), and we don't have to check all over the place.

Reply:
I applied the passing it as parameter everywhere, but don't like it, because it has a pertty big fallout radius.

It did spot two places where there wasn't checked for nullability yet though, so I guess it's better this way. I'll commit it and let you decide if you can allow the ugliness of passing it in almost every method.